aten/src/ATen/core/type.cpp

```diff
@@ -747,6 +751,74 @@ TupleTypePtr TupleType::createNamed(
       field_types, qualName, schema)); // NOLINT(modernize-make-shared)
 }
 
+void flatten_union(TypePtr& type, std::vector<TypePtr>& res) {
+  if (type->kind() == UnionType::Kind) {
+    for (auto inner_type : type->expect<UnionType>()->types()) {
+      flatten_union(inner_type, res);
+    }
+  }
+  else {
+    res.emplace_back(type);
+  }
+}
+
+UnionType::UnionType(std::vector<TypePtr> types) :
+  Type(TypeKind::UnionType) {
+  // Flatten Unions of Unions while populating `types_`
+  for (auto type : types) {
+    flatten_union(type, types_);
+  }
+  // We want the elements to be sorted so we can easily compare two UnionType
+  // objects for equality in the future
+  std::sort(types_.begin(), types_.end(),
+  [](const TypePtr a, const TypePtr b) -> bool {
+    return a->kind() > b->kind();
+  });
+  // Filter out duplicate types
+  types_.erase(std::unique(types_.begin(), types_.end()), types_.end());
+  // Check if this Union allows None
+  can_hold_none_ = std::any_of(types_.begin(), types_.end(),
+    [](const TypePtr type) { return type->kind() == NoneType::Kind; });
+}
+
+bool UnionType::isSubtypeOfExt(const TypePtr& rhs_, std::ostream* why_not) const {
+  if (auto union_rhs = rhs_->cast<UnionType>()) {
+    return std::all_of(this->types_.begin(), this->types_.end(),
+      [&](TypePtr this_type) -> bool {
+        return union_rhs->can_hold_type(this_type);
+    });
+  }
+  if (auto optional_rhs = rhs_->cast<OptionalType>()) {
+    return types_.size() == 2
+      && can_hold_type(optional_rhs->getElementType())
+      && can_hold_none();
+  }
+  if (Type::isSubtypeOfExt(rhs_, why_not)) {
+    return true;
+  }
+  if (rhs_->kind() == AnyTupleType::Kind) {
+    return true;
+  }
+  return false;
+}
+
+std::string UnionType::str() const {
+  std::stringstream ss;
+  ss << "Union(";
```

Comment:
I am a little surprised that this works since `str()` is supposed to return a legal type annotation, so in theory it should be `Union[`, what am I missing? Same for the `?` print on line 817

Reply:
I thought that `annotation_str` and `str` were used in different places and needed to be specified differently. Here, I'm following the example in `DictType` (for the parentheses) and `OptionalType` (the question mark representing a contained `None`).

Reply:
I think `annotation_str` is supposed to be valid Python type annotation, while `repr_str` is supposed to be human readable type printing used in error messages. `str` is the default implementation for both.

https://github.com/pytorch/pytorch/blob/19a0eb4cdbbe2d960c2b314352bf28b9b3cb9e9d/aten/src/ATen/core/jit_type_base.h#L74

So I think it makes sense to conform to Python's type annotation syntax, aka `Union[T1, T2, ...]`, let me know what you think.

Reply:
Thank you!